src/loaders/compute_tools/tool_common.py

```diff
@@ -269,6 +270,9 @@ def parallel_single_execution(self, tool_callable: Callable[[str, Path, Path, bo
                           meta[loader_common_names.META_SOURCE_FILE]),
                  output_dir,
                  self._debug))
+        
+        _create_metadata_file(genomes_meta, batch_dir)
```

Comment:
Can you explain why this needed to be moved from below? If there's an error in the execution stage that means the metadata file could exist without the tool output. I'm not sure what effect that might have.

Reply:
Oh, maybe it's so the file exists when the tool tries to parse it. In that case I would
* Move the metadata file creation back to where it was
* Pass the `kbase_id` into the tools. My first thought would be to change the first argument to a list of named tuples with the `genome_id`, the `kbase_id`, and the file path

Reply:
https://github.com/kbase/collections/pull/329#discussion_r1228619440

We kinda agreed upon to get kbase_id from meta file. And meta file must exist before we can parse them. If that's the case, we need to move metafile create step before [multiprocessing self.excute step](https://github.com/kbase/collections/blob/0b4f64ee1583038d60978508d93222b58a620dde/src/loaders/compute_tools/tool_common.py#LL277C1-L277C1)

Reply:
Well, that's the solution I naively proposed then without realizing that the metadata file didn't exist at that step. Since that's not the case, we should do something else rather than create a potential race condition; plus it makes the API more explicit if we pass in what we need rather than writing a file and having the code read it.

Reply:
Can you explain why there is a race condition here? A metafile is written to disk per batch_dir before any multiprocessing step is triggered though.

Reply:
Because the metadata file signals to subsequent steps that the processing is done (even if currently the processing steps don't throw an exception if the metadata file is missing - they probably should if not). If a tool does a partial write the next step will see a metadata file and assume the tool completed successfully.

Regardless of that, there's zero reason not to just pass in the necessary data to the tool processing code explicitly rather than passing in some of it and writing some of it to a file. That's just confusing.

Reply:
genome_id: "69036_857_1_kbase" --> kbase_id: "69036_857_1"
Are we falling back to certain structure/pattern?  https://github.com/kbase/collections/blob/a1c4012e885b8a26ce1a9fd3ab159766014274a3/src/loaders/compute_tools/tool_common.py#L51

Reply:
Actually, it's not a race condition - it can't be since there's no concurrency here. I tricked myself due to the out of order operations. Not sure what it'd be called, ambiguous termination maybe?

Anyway, the reasons we should avoid it are the same

Reply:
Re the kbase ID - I don't actually remember why we're appending that suffix - @Tianhao-Gu?

But anyway, we should just treat the genome_Id as an opaque string and the mapping we create in `ToolRunner` (which is written to the metadata file) as the canonical way to translate `genome_id`<->`kbase_id`. That way if we want to change the suffix, need to MD5, etc. the `genome_id` in the future no other code changes will be needed

Reply:
I assume we are still taking the mapping approach for now. Anything else I need to fix for this conversation?

Reply:
Not sure what you mean by the mapping approach

Reply:
Anything else we need to do? Current codebase is already using the metadata file to map genome_id<->kbase_id.

Reply:
Yes, we should put the metadata file creation back to after processing is done and pass in the kbase_id to the tool method instead

Reply:
Sounds logical so that we don't need the mapping to do the conversion genome_id<->kbase_id, but why we are appending that suffix to begin with? @Tianhao-Gu

Reply:
for GTDB genomes, GTDB-TK sometimes fails due to same reference genome id. That's why we append `_kbase` to the genome_id when run GTDB-TK tool.

Reply:
Hi, I put the metadata file creation back to after processing is done and passed in the genome_id<->kbase_id mapping into batch_execution and kbase_id into single_execution. Also tested in all tools for ENIGMA data.

Reply:
Updated!

Reply:
Rather than adding another argument just expand the first argument from a genome_id -> path mapping to either a list of named tuples or a mapping of genome_id -> NamedTuple[Path, kbase_id] or something like that.

Also, the single batch execution method should pass in the gneome_id as it did originally, but here we should add an argument for the kbase_id

Reply:
:+1: